src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java

```diff
@@ -605,16 +612,35 @@ public HttpResponse onSuccess(String authorizationCode) {
         };
     }
     
-    public void updateAuthorizationStrategy(OAuthSession oauth) throws IOException, GeneralSecurityException {
+    public UsernamePasswordAuthenticationToken updateAuthorizationStrategy(Credential credential) throws IOException, GeneralSecurityException {
     	populateDefaults();
-        OpenShiftUserInfo info = getOpenShiftUserInfo(oauth.getCredential(), transport);
-        Set<String> allowedRoles = postSAR(oauth.getCredential(), transport);
+        OpenShiftUserInfo info = getOpenShiftUserInfo(credential, transport);
+        Set<String> allowedRoles = postSAR(credential, transport);
         GrantedAuthority[] authorities = new GrantedAuthority[] {SecurityRealm.AUTHENTICATED_AUTHORITY};
+                
+        // we append the role suffix to the name stored into Jenkins, since a given user is able to log in at varying scope/permission
+        // levels in openshift; however, for now, we make sure the display name for Jenkins does not include this suffix
+        String suffix = "";
+        if (allowedRoles.contains("admin"))
+            suffix = "-admin";
+        else if (allowedRoles.contains("edit"))
+            suffix = "-edit";
+        else if (allowedRoles.contains("view"))
+            suffix = "view";
+        String matrixKey = info.getName() + suffix;
+        
         // logs this user in.
         UsernamePasswordAuthenticationToken token =
-                new UsernamePasswordAuthenticationToken(info.getName(), "", authorities);
+                new UsernamePasswordAuthenticationToken(matrixKey, "", authorities);
         SecurityContextHolder.getContext().setAuthentication(token);
         
+        User u = User.get(token.getName());
+        info.updateProfile(u);
```

Comment:
"userinfo" is essentially the "ping" ... it accesses the "/oapi/v1/users/~" URI on the master and sees if a user can read his own user info ... i.e. the user even exists.  If not, we don't either bother with the self-sar requests.

the "profile" object on the jenkins User object is just a means for updating metadata.  It provides access to any emails for example that the back end store (openshift oauth in this case) provides for the user.

The real key is these lines:

```
        UsernamePasswordAuthenticationToken token =
                new UsernamePasswordAuthenticationToken(matrixKey, "", authorities);
        SecurityContextHolder.getContext().setAuthentication(token);
```

But based on trying to explain this, I did find one flaw (I don't *think* it is exactly what you've been citing, but it is in the same ballparck).  I'll update the code with the fix I think it needs, plus some comments.  Hopefully that both shores things up and makes it clearer to observers to make sense of things.